[김성주] sprint7

626-ju opened the pull request:
https://panda-sprint7.netlify.app/items/199

## 요구사항

### 기본
#### 상품 상세
- [x]  상품 상세 페이지 주소는 "/items/{productId}" 입니다.
- [x]  response 로 받은 아래의 데이터로 화면을 구현합니다.
   - favoriteCount : 하트 개수
   - images : 상품 이미지
   - tags : 상품 태그
   - name : 상품 이름
   - description : 상품 설명
- [x]  목록으로 돌아가기 버튼을 클릭하면 중고마켓 페이지 주소인 "/items" 으로 이동합니다.

#### 상품 문의 댓글
- [x]  문의하기에 내용을 입력하면 등록 버튼의 색상은 "3692FF"로 변합니다.
- [x]  response 로 받은 아래의 데이터로 화면을 구현합니다.
   - image : 작성자 이미지
   - content : 작성자가 남긴 문구
   - description : 상품 설명
   - updatedAt : 문의글 마지막 업데이트 시간

### 심화

- [x]  모든 버튼에 자유롭게 Hover효과를 적용하세요.


## 주요 변경사항

-
-

## 스크린샷

#### pc
![pc](https://github.com/user-attachments/assets/187edc90-f247-4182-a636-fbb4b801f014)

#### 태블릿
![태블릿](https://github.com/user-attachments/assets/49584b74-42d9-417c-976a-9991bb63edb4)

#### 모바일
![모바일](https://github.com/user-attachments/assets/c9806041-8628-4703-b65c-10e2f284bbec)

#### 댓글 입력 시 버튼 활성화
![댓글 입력시 버튼 활성화](https://github.com/user-attachments/assets/ba1a6886-2141-4171-bea0-24667da089ba)

####  댓글 없을 때
![댓글없을 때](https://github.com/user-attachments/assets/e9d1ecd1-965a-45a0-8fdf-b4f2c8c08c64)


#### 댓글 수정 시
![댓글 수정 시](https://github.com/user-attachments/assets/8b009a9d-3ac3-4b28-b30a-e48a6ecd9e22)


## 멘토에게

-
-
- 셀프 코드 리뷰를 통해 질문 이어가겠습니다.

626-ju committed:
feat:미션7ui

626-ju committed:
feat:케밥 추가

626-ju committed:
chore: KebabMenu로 이름 변경

Commit:
feat: 수정하기 클릭 시 공용 폼 사용하기

626-ju committed:
fix: 백링크 경로 수정

626-ju committed:
feat:케밥 버튼들 구현 안한 api요청 alert로 막기

626-ju committed:
feat: button들 호버 효과

626-ju committed:
style: 스타일 수정

626-ju committed:
style: css 수정

626-ju committed:
feat: 댓글 작성 시간 상대시간변환 함수 추가

626-ju committed:
feat: 대체이미지  추가

626-ju committed:
feat: 댓글 스크롤 시 추가 패칭

626-ju committed:
refactor: 조건부렌더링->객체에 위임하기

626-ju committed:
fix: intersect트리거 ref 옮기기

626-ju committed:
chore: ProductInfo.jsx  article->section으로 변경

626-ju committed:
style: 포멧팅 수정

626-ju committed:
style: 포멧팅 수정

626-ju committed:
refactor: 에러핸들링 추가

626-ju committed:
style:포멧팅 수정

626-ju committed:
ProductItem.jsx getItem->showProductDetail 함수이름 변경

addiescode-sj commented on src/components/KebabMenu.jsx:
```diff
@@ -0,0 +1,64 @@
+}
+
+function KebabMenu({
+  dataList = ['수정하기', '삭제하기'],
```
해당 props값이 변하지않는 값이고, 이후 외부에 의한 변경을 원치않는다면 props보다는 상수로 관리해보는게 어떨까요? 너무 미래의 변경/수정 가능성을 생각해 개발하기보다는, 현재에 집중하는게 좋아요 :)

addiescode-sj commented on src/components/UserProfileCard.jsx:
```diff
@@ -0,0 +1,25 @@
+};
+
+function UserProfileCard({ authority, name, time, image }) {
+  const userImage = image ? image : profileImage;
```
```suggestion
  const userImage = image || profileImage;
```

이렇게 좀 더 단순하게 쓸 수 있겠네요!

addiescode-sj commented on src/components/UserProfileCard.jsx:
```diff
@@ -0,0 +1,25 @@
+import profileImage from '@assets/icon/ic_user.png';
+import styles from '../styles/UserProfilCard.module.css';
+
+const classNameByauthority = {
+  post: styles.post,
+  comment: styles.comment,
+};
```
클래스이름을 authority props value에 의해 조합하는군요! 굳굳 👍 
일관성을 위해 이렇게 조합하는 방식이 이후 여러가지로 많아진다면 이런 라이브러리를 도입해보는건 어떨까요?

https://www.npmjs.com/package/clsx

addiescode-sj commented on src/hooks/useLoadMoreComments.js:
```diff
@@ -0,0 +1,22 @@
+import { useEffect, useRef } from 'react';
+
+export function useLoadMoreComments(targetRef, onIntersect = () => {}) {
```
해당 훅은 직접적으로 코멘트를 로드한다는 관심사를 가지고 있지 않은걸로 보이네요!
`useIntersectionObserver` 정도로 네이밍하는게 적당해보입니다 :)

addiescode-sj commented on src/pages/ProductDetail/AskForm.jsx:
지금도 가독성에 전혀 문제가 없고 좋지만, 더 유지보수하기 좋은 코드로 만들기 위해 해당 파일을 전체적으로 리팩토링해볼까요?

우선 첫번째로 파일에서 필요한 상수가 어떤것이고, 해당 상수에 따른 스타일과 분기를 좀 더 명확히 분리해 사용해보도록 합시다. 

`formHeightByMethod`와 `buttonByMethod`에서 공통적으로 POST, UPDATE 상수가 중복되어있네요 :)

```jsx
// Constants
const METHODS = {
  POST: 'POST',
  UPDATE: 'UPDATE',
};
```

addiescode-sj replied:
이렇게 공통된 상수를 명확히 구분해주면,
해당 상수에 따른 스타일과 분기 또한 아래처럼 명확히 분리해줄수있는 이점이 있습니다.

```jsx
const FORM_HEIGHT_BY_METHOD = {
  [METHODS.POST]: { height: '129px' },
  [METHODS.UPDATE]: { height: '80px' },
};
```

```jsx
  function onSubmit(e) {
    if (method === METHODS.POST) alert('준비 중인 기능입니다(댓글 등록)');
    else if (method === METHODS.UPDATE) alert('준비 중인 기능입니다(댓글 수정)');
    e.preventDefault();
  }
```

addiescode-sj replied:
그리고, 버튼을 렌더하는 렌더러 함수도 공통으로 빼둔 상수를 사용해 이렇게 사용해보면 어떨까요?

```jsx
// Button renderer
const renderButton = (method, { notEmpty, setIsEditNow }) => {
  switch (method) {
    case METHODS.POST:
      return <PostBtn notEmpty={notEmpty} />;
    case METHODS.UPDATE:
      return <UpdateBtn notEmpty={notEmpty} setIsEditNow={setIsEditNow} />;
    default:
      return null;
  }
};
```

addiescode-sj replied:
어떤가요? 이전에 비해 공통으로 쓰이는 상수에 따른 조건별 처리가 명확해지니 읽기도 쉬워지고 확장성이 생기지않나요? :)

addiescode-sj commented on src/pages/ProductDetail/CommentItem.jsx:
이 파일에서도 바로 위에 드린 코멘트 참고해보시고, 리팩토링해보세요 :)
- 공통으로 쓰이는 상수를 분리하고, 조건별 처리를 명확하게 해보기

addiescode-sj reviewed (approved):
수고하셨습니다!
성주님 코드는 항상 읽기 편하네요 :)

## 주요 리뷰 포인트
- props값 사용 필요 고민해보기
- 일관성을 위해 클래스이름 조합 방식 통일해보기
- `useLoadMoreComments` 네이밍 개선
- 공통 상수 분리하고, 조건별 처리를 명확히 하는 리팩토링 방법 제시

626-ju committed:
refactor(mentor): ||로 기본값 설정

626-ju committed:
refactor(mentor): useLoadMoreComments->useIntersectionObserver로 이름 변경

626-ju committed:
refactor(mentor): Kebab props로 받는 드랍다운 리스트 항목 상수

626-ju committed:
refactor(mentor): 상수에 따라 분기 명확히 나타내기(AskForm.jsx)+

-Kebab에서 관리하던 DROPDOWN_LIST를 케밥 호출하는 곳에서 상수 객체로
관리
(ProductInfo.jsx, CommentItem.jsx)

626-ju committed:
fix: 린트오류수정 및 포맷팅 +

-ProductsAll.jsx, ProductsFavorite.jsx에서
아이템들 lazy로딩하던 거 함수 밖으로 빼기

626-ju committed:
refactor(Auth):Login.jsx 리액트 훅 폼 사용해보기

626-ju committed:
refactor(Auth): Login.jsx useForm에 mode초기값 all로 수정+

필드에서 내려주던 register,errors FormInput안에서 호출하기

626-ju committed:
refactor(Auth): SignUp.jsx도  react-hook-form사용해보기+

password 입력시 passwordCheck 트리거는 좀 더 고민해보기

626-ju committed:
style(Auth): auth-> login으로 클래스 이름 변경

626-ju committed:
refactor(Home): Footer.jsx snsList 컴포넌트로 분리

626-ju committed:
chore: 불필요한 주석 제거
